man/fi_peer.3.md

```diff
@@ -0,0 +1,162 @@
+of which method is employed.  However, because the goal of using peer
+providers is to avoid overhead, providers must be explicitly written to
+support the peer provider mechanisms.
+
+Peer providers are identified as either being an exporter or importer.
+A provider can only be an exporter or importer.  That is, a provider
+cannot export one object while importing another.  The exporter allocates
+and owns the fabric objects that are accessed by the libfabric user.
+Note that the exporter maintains the set of addresses of all communicating
+endpoint addresses.  An object may be imported by multiple providers.
+
+## fi_export_fid / fi_import_fid
+
+Generally, fabric objects are allocated and managed entirely by a single
+provider.  Typically only the application facing software interfaces of
+a fabric object are defined, for example, the message or tagged operations
+of an endpoint.  The fi_export_fid and fi_import_fid calls provide a
+a mechanism by which provider facing APIs may be accessed.  This allows
+the creation of fid objects that are shareable between providers, or
+for library plug-in services.  The ability to export a shareable object
+is object and provider implementation dependent.
+
+Shareable fids typically contain at least 3 main components: a
+base fid, a set of exporter defined ops, and a set of importer defined
+ops.
+
+# PEER CQ
+
+The peer CQ defines a mechanism by which a provider may insert completions
+into the CQ of a peer provider.  This avoids the overhead of the libfabric
+user from needing to access multiple CQs.
+
+To setup a peer CQ, one CQ must be identified as the CQ exporter.
+The exporter CQ is created through the standard fi_cq_open() call by setting
+the FI_PEER_EXPORT flag with the fi_cq_attr::flags field.  If the provider
+supports peer CQs, it must configure the CQ such that it will be accessible
+by a peer.  Providers that do not support peer CQs should fail the fi_cq_open
+call with an invalid flag (-FI_EINVAL).  Note that if the exporting provider
+is the entity setting up the peer providers, then it will simply assume the
+export role without the FI_PEER_EXPORT flag being set.
+
+All other peer CQs are identified as CQ importers.  Peer CQs are also
+configured using the fi_cq_open() call, but by passing in the FI_PEER_IMPORT
+flag.  When FI_PEER_IMPORT is specified, the fi_cq_open() call is modified as
+described.  The cq parameter is treated as input/output.  On input, the cq
+of the exporter which was returned by fi_cq_open() is provided.  On output,
+the importer will return its own cq structure.  The importer's CQ should be
+configured to report completions through the exporter.  As a result, the
+CQ functions for reading the importer CQ (e.g. fi_cq_read, fi_cq_readerr,
+etc.) should be set to return -FI_ENOSYS when called.
+
+The importer is responsible for setting up the callbacks with the exporter.
+It does this by calling fi_export_fid(), passing in the input cq passed to
```

Comment:
The importer is exporting the fid from the peer.  The fi_import_fid / fi_export_fid APIs are already defined.  We need to use fi_export_fid() as that allows returning the new object.
```
fi_export_fid(struct fid *fid, uint64_t flags,
	      struct fid **expfid, void *context);
fi_import_fid(struct fid *fid, struct fid *expfid, uint64_t flags)
```
We could create a new wrapper, say, fi_get_peer_fid()

Reply:
Yes. I also found it a little confusing

But `fi_import_fid` is being used by open-mpi:

https://github.com/open-mpi/ompi/blob/dfb9ca0bb5d253366512a5fabc11efae9c58f57c/opal/mca/common/ofi/common_ofi.c#L172

Reply:
A full sequence could explain it better. Here is my understanding:

exporter:
(1) call fi_cq_open() with FI_PEER_EXPORT flag, output cq_fid;
(2) call fi_export_fid() with cq_fid as input, output cq_exfid with export_ops set, also associate cq_fid with cq_exfid.

importer:
(1) call fi_cq_open() with FI_PEER_IMPORT flag, input cq_exfid, output cq_fid2. associate cq_fid2 with cq_exfid, set import_ops.

The importer needs not to know cq_fid from exporter. It should only deal with the exported fid.

Reply:
Alternatively, the importer should be able to call fi_import_fid() with cq_exfid as the input and local cq_fid as the output instead of calling fi_cq_open(). This match how fi_import_fid() is used for the memory monitor case.

Reply:
The latter lacks the ability to pass cq_attrs so is only applicable is all default setting are used.

Reply:
I think there need to be an `getattr` method, so importer can get the attribute of the CQ.

Reply:
Updated man page to reword things in places and add a call sequence.  I may update it again to define 3 roles - the exporter, the importer, and the linking entity.  But will wait for other comments.

Reply:
The importer is passed attributes through the fi_cq_open() call.  This way, the attributes of the importer can be independent from the attributes of the exporter.  It's hard to imaging how most of the attributes would be used by the importer, but I wanted them there.  At least the cq_attr::format is useful.  Knowing if the exporter is supporting blocking calls might be.

Something not captured by the changes, but that I considered, is that the importer may still need to handle the fi_trywait() call.  But that also gets into sharing fd's (or some other wait object) between the providers.  And since we're really only targeting the shm provider as a peer for now, I decided to ignore this problem.